### DISCUSSION TITLE
Data type archives organization

### DISCUSSION THREAD

**Discussion by Gravelbones** [NationalSecurityAgency/ghidra#5185] [Q&A]
I have been looking into the C parser and building data type archives for old DOS sources.
That have raised a number of question about how to organize data type archives and what is the proper way to add them to a project.

So this is effort to try to describe these question and the background.
I am looking for indication of what way the main project is moving, but also how others might see this.
Don't take anything literally but as an example of what might be the case.
Don't tell how things are now, but rather what would be the optimal solution.

## Super archives
Currently there is a set of windows super archive windows_vs12*, a set of generic_clib, and a single mac_osx.
The code for selecting one of them is in the data manager util code.
But that seems very limited and most loader formats end up with generic_clib.

So where/how to add DOS (or any other OS), or other languages for Windows (like Pascal)?
Should there be "tools" to maintain such super archives from multiple sources?
Should the selection of default data type archive be the job of the loader (a function to return default datatype archive)?

I understand why the include files need to be parsed for both 32 and 64 bit.
But shouldn't the resulting archive be able to handle both results without having to do a different type?
Or is it a part of the version problem below?

## Contribution support
If a contributor (like me) would like to add a larger number of datatypes to a super archive,
what format is preferable: GDT, include file, JSON, or some XML?
What if the include file is not public domain, like Borland may still have copyright of their include files?
Should those be restricted (barred) as input to the Ghidra project regardless of type?
Not that anyone may be likely to pursue exactly that case.
Should there be a project of its own, trying to gather the include files there?

## Multiple parses
As at least one issue indicate, parsing the files only once may not be enough to get the various definitions.
Also compiler switches may add different defines, requiring a parse all of its own.
Doing a output of defines for each compiler, will list many more than those defined by Ghidra parsning.
But how far should it be taken?
If it can be automated or added to a generation script, should most possible combinations be added?
Should it be possible to recreate a data type archive from scratch (how to maintain the unique ID)?

Unlike a given program when being compiled, Ghidra needs to know every variation a given include file set may result in.

## Versions of structures
Between multiple version of C compilers, different version of the OS and so on, a given structure may be different.
For example st_stat from fstat. There is not one standard definition of this, and not a given order the fields should be in.
So how should these multiple version be added to a data type archive and how should they be selected?
Also between versions of DOS some structures fields are used differently, even when the structure remains the same size.
Should these structures be turned into unions or be completely different versions?
And how about the different versions like directdraw 9/10/11/12, which are all available in the same windows AFAIK?
Not that I have checked if all names are kept different between the versions and therefore there is no problem.
How to select the correct version from an analyzer?

And how about libraries like openssl, which can be used across platforms, or soundblaster libraries for DOS and
maybe even windows if working with a device driver.
Should a data type archive be created for each or should there be some kind of "super" archive for support libraries?
Or should they simply be part of the super archives?

What if a structure has a fixed size across platforms and it is defined with int of varying numbers.
Should those be detected during multiple parses?
Or is it just another variation of version problem above.

## Use Enums more
When compiling it is easy to convert from a constant name to an integer value.
But when Ghidra needs to turn a integer value into a constant name, there are many to choose from.
But by adding extra information when parsing include files, Ghidra can be given some more information, than what appears in the include file.
An example

- fseek(FILE *f, int, int) or
- fseek(FILE *f, int, off_t)

The second integer should be a value of 0 - 2 with the constant names SEEK_SET, SEEK_CUR or SEEK_END.
But Ghidra has no way to do this automatically. But if the definition was changed to

- fseek(FILE *f, seek_enum, off_t)

and seek_enum was a enum type, with the given values, Ghidra would be able to map a integer value to a constant name
automatically, as far as I have seen.
But is this a way to go?

Also using the various off_t, size_t, time_t rather than int, some assumption about the type can be done.
Should a parse of newer files, help with changing these types of older functions?

A special case could be the need to extend enums from one archive with values from another archive.
Windows command enum comes to mind. But I will assume this should be handled dynamically and not be a part of this.

## Comments
Emteere made the suggestion/goal of using the comments inside the include file to augment the structures created.
Should such an approach go toward a relaxed position, creating possible wrong comments.
Or should it be strict only adding comments it is certain is correct?
If information support needed for enums above are implemented, I would expect the same should support extraction of comments/adding missing comments.

I have seen cases where comments come before and others where it comes after.
So no easy way to do this.

## 8086 real mode pointers
Another problem for real mode DOS and Windows programs are the different function versions for near/far pointers.
There is a need for having both

- 	strcpy(char near *, char near *)
- 	strcpy(char far *, char far *)

where the size of the pointer is either 2 or 4 bytes, but not based on processor.
How should the different versions be saved?
How to select the correct version from an analyzer (same problem with versions of structs)?
For now I only care about the problem of function types, not the individual pointers.
Supporting the base near/far/segment type is another problem saved for a later stage.

## Language dialect support
Even C and C++ have many variations.
The current C parse is trying to support those by being relaxed in its parsing.
Going forward should this still be the goal, or if possible should it be more strict?
Personally I will go for relaxed as that will be easier for most users.
Also the current C parser seem to have some very limited Objective C support, but I have no test cases or idea of
how the result should be.
Should the same parser be able to support versions which are quite different or do the user know that much, and
can choose the proper "dialect"?

## XSYM
Also I have found no examples of the XSYM files that are supported.
Can XSYM be "connected" to different versions of C?

## Object oriented languages
This may extend into the realm of object oriented languages (C++).
Is there any work going in representing class, templates and so on?
How about the virtual table for a class, does these differ across compilers?
Is there any plan/wish on doing a plugin for template based construction of classes from templates?

## What have I forgotten?
What elements have I forgotten in the big picture.
I know there are a lot of details to sort out.

**Comment by emteere** [comment]
All good questions and there is a lot of them to answer, so it may take a bit to dig through to provide a good answer.  There are definitely changes coming in the area of data types to better handle architecture, data organization, beyond 32/64 that will have an affect on your questions.

Some of the questions you are asking are a bit intractable, but may have some potential solutions, possible directions for Ghidra, or techniques to deal with them.

I've read through your questions and I'll try to address what I can another day.

**Reply by Gravelbones** [reply]
Take your time. I will do work that is not depending on any of these questions. And won't have time for the next week at least anyway.
And I don't expect you to provide solutions to all my answers. Some are indeed intractable, but just say so.
I will gladly offer my ideas and solutions. Or leave it to the user for the time being.
